express/parsers/apps/espresso/formats/espresso_640xml.py

```diff
@@ -59,6 +61,79 @@ def nspins(self):
 
         return nspin
 
+    def eigenvalues_at_kpoints(self):
+        """
+        Returns eigenvalues for all kpoints.
+
+        Returns:
+             list
+
+        Example:
+            [
+                {
+                    'kpoint': [-0.5, 0.5, 0.5],
+                    'weight': 9.5238095E-002,
+                    'eigenvalues': [
+                        {
+                            'energies': [-1.4498446E-001, ..., 4.6507387E-001],
+                            'occupations': [1, ... , 0],
+                            'spin': 0.5
+                        }
+                    ]
+                },
+                ...
+            ]
+        """
+
+        # Determine if we have multiple spin states
+        # We're doing it this way to avoid coupling to the `self.nspins()` behavior. If there are up/down bands, then
+        # we know there are multiple spins. Otherwise, it only makes sense to treet it as having a single spin state.
+        bandstructure_node = self.traverse_xml(self.root, ("output", "band_structure"))
+        if bandstructure_node.find("nbnd_up"):
+            has_multiple_bands = True
+        else:
+            has_multiple_bands = False
+
+        kpoints = []
+        for ks_energy_node in bandstructure_node.findall("ks_energies"):
+            # Basic information about the kpoint
+            kpoint = {}
+            kpoint_node = ks_energy_node.find("k_point")
+            kpoint["kpoint"] = string_to_vec(kpoint_node.text)  # Coordinates
+            kpoint["weight"] = float(kpoint_node.get("weight"))
+
+            # Extract eigenvalues
+            eigenvalue_text = ks_energy_node.find("eigenvalues").text
+            occupation_text = ks_energy_node.find("occupations").text
+            energies = [component * Constant.HARTREE for component in string_to_vec(eigenvalue_text)]
+            occupations = string_to_vec(occupation_text, dtype=float)
+
+            # Split into up/down spin if we need to
+            # In the case of having multiple spins, the spin up states are listed before the spin down states, hence
+            # the list slicing.
+            if has_multiple_bands:
+                nband_up = int(bandstructure_node.find("nbnd_up").text)
+                nbnd_down = int(bandstructure_node.find("nbnd_dwn").text)
```

Comment:
I think `nbnd_down` is unused currently

Reply:
That's a good point. I'd included it just so it's there, but we aren't really using it